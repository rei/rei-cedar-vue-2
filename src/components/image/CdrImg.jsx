import clsx from 'clsx';
import modifier from '../../mixins/modifier';
import style from './styles/CdrImg.scss';

export default {
  name: 'CdrImg',
  mixins: [modifier],
  props: {
    /**
     * Required. Image source url.
     */
    src: {
      type: String,
      required: true,
    },
    /**
     * Required. Image alt text.
     */
    alt: {
      type: String,
      default: '',
    },
    /**
     * Enable lazy loading.
     */
    lazy: {
      type: Boolean,
    },
    /**
     * Object of lazy options
     */
    lazyOpts: {
      type: Object,
      default: () => {},
    },
    /**
     * Aspect ratio of the media container. {auto, square, 1-2, 2-3, 3-4, 9-16, 2-1, 3-2, 4-3, 16-9}
     */
    ratio: {
      type: String,
      validator: value => ([
        'auto',
        'square',
        '1-2',
        '2-3',
        '3-4',
        '9-16',
        '2-1',
        '3-2',
        '4-3',
        '16-9'].indexOf(value) >= 0) || false,
    },
    /**
     * sm breakpoint and above
     */
    ratioSm: {
      type: String,
      validator: value => ([
        'auto', 'square', '1-2', '2-3', '3-4', '9-16', '2-1', '3-2', '4-3',
        '16-9'].indexOf(value) >= 0) || false,
    },
    /**
     * md breakpoint and above
     */
    ratioMd: {
      type: String,
      validator: value => ([
        'auto', 'square', '1-2', '2-3', '3-4', '9-16', '2-1', '3-2', '4-3',
        '16-9'].indexOf(value) >= 0) || false,
    },
    /**
     * lg breakpoint and above
     */
    ratioLg: {
      type: String,
      validator: value => ([
        'auto', 'square', '1-2', '2-3', '3-4', '9-16', '2-1', '3-2', '4-3',
        '16-9'].indexOf(value) >= 0) || false,
    },
    /**
     * Requires a `ratio`. Area to crop the image overflow to. {top, y-center, bottom} {left, x-center, right}
     */
    crop: {
      type: String,
    },
    /**
     * Requires a `ratio`. Scale the image to be as large as possible to fill the area (background-position: cover;)
     */
    cover: {
      type: Boolean,
    },
    /**
     * Sets a border radius to an element {square, top, right, bottom, left}
     */
    radius: {
      type: String,
      validator: value => ([
        'circle',
        'rounded'].indexOf(value) >= 0) || false,
    },
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-image';
    },
    lazyClass() {
      const classObj = {};
      classObj['lazy-image'] = this.lazy;
      return classObj;
    },
    radiusClass() {
      const classObj = {};
      classObj[this.style[`cdr-image--${this.radius}`]] = this.radius;
      return classObj;
    },
    ratioClass() {
      const classObj = {};
      classObj[this.style[`cdr-media-frame--${this.ratio}`]] = this.ratio;
      classObj[this.style[`cdr-media-frame--${this.ratioSm}@sm`]] = this.ratioSm;
      classObj[this.style[`cdr-media-frame--${this.ratioMd}@md`]] = this.ratioMd;
      classObj[this.style[`cdr-media-frame--${this.ratioLg}@lg`]] = this.ratioLg;
      return classObj;
    },
    coverClass() {
      const classObj = {};
      classObj[this.style['cdr-media-frame__cover']] = true;
      classObj[this.style['cdr-media-frame__cover--crop']] = this.crop;
      classObj[this.style['cdr-media-frame__cover--cover']] = this.cover;
      return classObj;
    },
    cropClass() {
      const base = 'cdr-media-frame';
      const cropArr = this.crop ? this.crop.split(' ') : [];
      let final = [];

      final = final.concat(cropArr.map(mod => this.modifyClassName(base, mod)));

      return final.join(' ');
    },
    styleObject() {
      return {
        backgroundImage: `url(${this.src})`,
      };
    },
    lazyAttrs() {
      const attrObj = {};
      if (this.lazy) {
        Object.keys(this.lazyOpts).forEach((opt) => {
          attrObj[`data-${opt}`] = this.lazyOpts[opt];
        });
      }
      return attrObj;
    },
  },
  render() {
    if (this.ratio) {
      return (
        <div
          class={clsx(this.style['cdr-media-frame'], this.ratioClass, this.cropClass)}
        >
          <div
            class={clsx(this.coverClass, this.lazyClass, this.radiusClass)}
            style={this.styleObject}
            aria-hidden="true"
            {...{ attrs: this.lazyAttrs }}
          />
          <img
            class={clsx(
              this.style['cdr-media-frame__image'],
              this.style['cdr-media-frame__image--hidden'],
              this.modifierClass,
              this.radiusClass,
            )}
            src={this.src}
            alt={this.alt}
          />
        </div>
      );
    }
    return (<img
          class={clsx(this.modifierClass, this.radiusClass, this.lazyClass)}
          src={this.src}
          alt={this.alt}
          {...{ attrs: this.lazyAttrs }}
        />);
  },
};
