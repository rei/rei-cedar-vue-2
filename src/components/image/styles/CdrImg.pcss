/* ==========================================================================
  # CdrImg
  
  All values should map to variables in
  vars/CdrImg.vars.pcss
  in order to allow for theming

  TOC:

    :Base - Image
      :Modifiers
        :Style Variants
    :Base - Media Frame
      :Modifiers
        :Aspect Ratios
          :normal
          :@small
          :@medium
          :@large
        :Cropping
      :Elements
========================================================================== */

/* ==========================================================================
   # Image
   ========================================================================== */

.cdr-image {
  max-width: 100%;
  vertical-align: middle;
  border-style: none;

  /* Style variants
    ========================================================================== */

  /* responsive
    ========== */
  &--responsive {
    display: block;
    height: auto;
    width: 100%;
  }

  /* rounded
    ========== */

  &--rounded {
    border-radius: $radius-softer;
  }

  /* circle
    ========== */

  &--circle {
    border-radius: $radius-round;
  }
}

/* ==========================================================================
   # MEDIA FRAME
   ========================================================================== */

.cdr-media-frame {
  position: relative;
  overflow: hidden;

  &::before {
    content: '';
    height: 0;
    display: block;
  }

  /* Aspect ratios
    ========================================================================== */

  &--auto::before {
    padding-bottom: 0;
  }

  &--square::before {
    padding-bottom: 100%;
  }

  &--1-2::before {
    padding-bottom: 200%;
  }

  &--2-3::before {
    padding-bottom: 150%;
  }

  &--3-4::before {
    padding-bottom: 133.33333333333%;
  }

  &--9-16::before {
    padding-bottom: 177.77777777778%;
  }

  &--2-1::before {
    padding-bottom: 50%;
  }

  &--3-2::before {
    padding-bottom: 66.6666666667%;
  }

  &--4-3::before {
    padding-bottom: 75%;
  }

  &--16-9::before {
    padding-bottom: 56.25%;
  }

  /* Aspect ratios @ small
    ========== */
  @media (--sm-mq) {
    &--auto\@sm::before {
      padding-bottom: 0;
    }

    &--square\@sm::before {
      padding-bottom: 100%;
    }

    &--1-2\@sm::before {
      padding-bottom: 200%;
    }

    &--2-3\@sm::before {
      padding-bottom: 150%;
    }

    &--3-4\@sm::before {
      padding-bottom: 133.33333333333%;
    }

    &--9-16\@sm::before {
      padding-bottom: 177.77777777778%;
    }

    &--2-1\@sm::before {
      padding-bottom: 50%;
    }

    &--3-2\@sm::before {
      padding-bottom: 66.6666666667%;
    }

    &--4-3\@sm::before {
      padding-bottom: 75%;
    }

    &--16-9\@sm::before {
      padding-bottom: 56.25%;
    }
  }

  /* Aspect ratios @ medium
    ========== */
  @media (--md-mq) {
    &--auto\@md::before {
      padding-bottom: 0;
    }

    &--square\@md::before {
      padding-bottom: 100%;
    }

    &--1-2\@md::before {
      padding-bottom: 200%;
    }

    &--2-3\@md::before {
      padding-bottom: 150%;
    }

    &--3-4\@md::before {
      padding-bottom: 133.33333333333%;
    }

    &--9-16\@md::before {
      padding-bottom: 177.77777777778%;
    }

    &--2-1\@md::before {
      padding-bottom: 50%;
    }

    &--3-2\@md::before {
      padding-bottom: 66.6666666667%;
    }

    &--4-3\@md::before {
      padding-bottom: 75%;
    }

    &--16-9\@md::before {
      padding-bottom: 56.25%;
    }
  }

  /* Aspect ratios @ large
    ========== */
  @media (--lg-mq) {
    &--auto\@lg::before {
      padding-bottom: 0;
    }

    &--square\@lg::before {
      padding-bottom: 100%;
    }

    &--1-2\@lg::before {
      padding-bottom: 200%;
    }

    &--2-3\@lg::before {
      padding-bottom: 150%;
    }

    &--3-4\@lg::before {
      padding-bottom: 133.33333333333%;
    }

    &--9-16\@lg::before {
      padding-bottom: 177.77777777778%;
    }

    &--2-1\@lg::before {
      padding-bottom: 50%;
    }

    &--3-2\@lg::before {
      padding-bottom: 66.6666666667%;
    }

    &--4-3\@lg::before {
      padding-bottom: 75%;
    }

    &--16-9\@lg::before {
      padding-bottom: 56.25%;
    }
  }

  /* Cropping variants
    ========================================================================== */

  &--x-center > &__cover {
    background-position-x: center;
  }

  &--y-center > &__cover {
    background-position-y: center;
  }

  &--left > &__cover {
    background-position-x: left;
  }

  &--right > &__cover {
    background-position-x: right;
  }

  &--top > &__cover {
    background-position-y: top;
  }

  &--bottom > &__cover {
    background-position-y: bottom;
  }

  /* Block elements
    ========================================================================== */

  &__image {
    z-index: -1;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;

    /* Hide image when only used for a11y */
    &--hidden {
      opacity: 0;
    }
  }

  &__cover {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    background-size: contain;
    background-repeat: no-repeat;
    background-position: center;

    &--crop {
      background-size: auto;
    }

    &--cover {
      background-size: cover;
    }
  }
}
