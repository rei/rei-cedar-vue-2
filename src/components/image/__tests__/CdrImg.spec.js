import { shallowMount, mount } from '@vue/test-utils';
import CdrImg from 'componentdir/image/CdrImg';

describe('CdrImg', () => {
  test('renders correctly', () => {
    const wrapper = mount(CdrImg, {
      propsData: {
        src: 'http://via.placeholder.com/350x150',
        alt: 'test alt',
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders crop/ratio correctly', () => {
    const wrapper = mount(CdrImg, {
      propsData: {
        ratio: "square",
        cover: true,
        crop: "left",
        alt: "crop left",
        src: "http://placehold.it/1920x1080",
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('sets the alt attr correctly', () => {
    const wrapper = shallowMount(CdrImg, {
      propsData: {
        src: 'http://via.placeholder.com/350x150',
        alt: 'test alt',
      }
    });
    expect(wrapper.attributes().alt).toBe('test alt');
  });

  it('sets the alt attr correctly in a media frame', () => {
    const wrapper = shallowMount(CdrImg, {
      propsData: {
        src: 'http://via.placeholder.com/350x150',
        ratio: 'square',
        alt: 'test alt',
      }
    });
    expect(wrapper.find('img').attributes().alt).toBe('test alt');
  });

  it('adds lazy class and attrs', () => {
    const wrapper = shallowMount(CdrImg, {
      propsData: {
        src: 'http://via.placeholder.com/350x150',
        lazy: true,
        lazyOpts: {
          'src-lazy': 'http://via.placeholder.com/350',
        },
      }
    });
    expect(wrapper.classes()).toContain('lazy-image');
    expect(wrapper.attributes()['data-src-lazy']).toBe('http://via.placeholder.com/350');
  });

});