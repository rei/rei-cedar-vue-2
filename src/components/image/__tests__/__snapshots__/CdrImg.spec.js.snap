// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrImg renders correctly 1`] = `
<img
  alt="test alt"
  class="cdr-image"
  src="http://via.placeholder.com/350x150"
/>
`;

exports[`CdrImg renders crop/ratio correctly 1`] = `
<div
  class="cdr-media-frame cdr-media-frame--square cdr-media-frame--left"
>
  <div
    aria-hidden="true"
    class="cdr-media-frame__cover cdr-media-frame__cover--crop cdr-media-frame__cover--cover"
    style="background-image: url(http://placehold.it/1920x1080);"
  />
  <img
    alt="crop left"
    class="cdr-media-frame__image cdr-media-frame__image--hidden cdr-image"
    src="http://placehold.it/1920x1080"
  />
</div>
`;
